src/Database/LSMTree/Internal/Lookup.hs

```diff
@@ -1,32 +1,165 @@
+      -- starting estimate based on the expected true- and false-positives.
+      res <- VUM.unsafeNew (2 * ksN)
+      resixVar <- newPrimVar 0
+      res' <- go res resixVar
+      resix <- readPrimVar resixVar
+      -- The result vector is not guaranteed to be full, so we slice it. This
+      -- means it might retain memory that is not used by the vector.
+      pure $ VUM.slice 0 resix res'
+  where
+    !rsN = V.length rs
+    !ksN = V.length ks
+
+    go :: forall s.
+         VUM.MVector s (RunIx, KeyIx)
+      -> PrimVar s ResIx
+      -> ST s (VUM.MVector s (RunIx, KeyIx))
+    go res0 resixVar = loop1 res0 0
+      where
+        -- Loop over all run indexes
+        loop1 ::
+             VUM.MVector s (RunIx, KeyIx)
+          -> RunIx
+          -> ST s (VUM.MVector s (RunIx, KeyIx))
+        loop1 !res1 !rix
+          | rix == rsN = pure res1
+          | otherwise
+          = do
+              res1' <- loop2 res1 0 (rlvBloom $ rs `V.unsafeIndex` rix)
+              loop1 res1' (rix+1)
+          where
+            -- Loop over all key indexes
+            loop2 ::
+                 VUM.MVector s (RunIx, KeyIx)
+              -> KeyIx
+              -> Bloom SerialisedKey
+              -> ST s (VUM.MVector s (RunIx, KeyIx))
+            loop2 !res2 !kix b
+              | kix == ksN = pure res2
+              | let !k = ks `V.unsafeIndex` kix
+              , Bloom.elem k b = do
+                  resix <- readPrimVar resixVar
+                  -- Grows the vector if we've reached the end. TODO: tune how
+                  -- much much we grow the vector each time based on the
+                  -- expected true- and false-positives.
+                  res2' <- if resix == VUM.length res2
+                            then VUM.unsafeGrow res2 ksN
+                            else pure res2
+                  VUM.unsafeWrite res2 resix (rix, kix)
+                  writePrimVar resixVar $! (resix +1)
+                  loop2 res2' (kix+1) b
+              | otherwise = loop2 res2 (kix+1) b
+
+-- | Perform a batch of fence pointer index searches, and create an 'IOOp' for
+-- each search result. The resulting vector has the same length as the
+-- @VU.Vector (RunIx, KeyIx)@ argument, because index searching always returns a
+-- positive search result.
+indexSearches ::
+     forall m h. PrimMonad m
+  => V.Vector (RunLookupView (Handle h))
+  -> V.Vector SerialisedKey
+  -> VU.Vector (RunIx, KeyIx) -- ^ Result of 'bloomQueries'
+  -> m (V.Vector (IOOp m h))
+indexSearches rs ks rkixs = do
+    -- The result vector has exactly the same length as @rkixs@.
+    res <- VM.unsafeNew n
+    go res
+    V.unsafeFreeze res
+  where
+    !n = VU.length rkixs
+
+    go :: VM.MVector (PrimState m) (IOOp m h) -> m ()
+    go res = loop 0
+      where
+        -- Loop over all indexes in @rkixs@
+        loop :: Int -> m ()
+        loop !i
+          | i == n = pure ()
+          | otherwise = do
```

Comment:
I guess you tried something like `V.mapM _ rkixs` instead of writing your own loop, but performance wasn't great? Is that where the vector bundle code showed up in core?

Reply:
I didn't actually try `mapM_` here, but it should be easy to change and see what happens to the performance

Reply:
Okay, looking at it again now it's actually probably not `mapM_` that we want. We're looping over the indexes into the result vector, not the elements in the result vector themselves

Reply:
But all you're doing with `i` is indexing into `rkixs` and writing to `res`, so I would say you're exactly mapping, just with more manual bookkeeping.

Reply:
Hehe you're right, my mistake

Reply:
Performance is nearly identical!